AndroidSDK/src/com/leanplum/ActionContext.java

```diff
@@ -410,47 +426,90 @@ private boolean isChainToExistingMessageStarted(Map<String, Object> args, String
               message.get(Constants.Keys.VARS));
           Object messageAction = message.get("action");
           return messageAction != null && createActionContextForMessageId(messageAction.toString(),
-              messageArgs, messageId, name);
+              messageArgs, messageId, name, true);
         }
       }
     }
     return false;
   }
 
   /**
-   * Prefix given event with all parent actionContext names to while filtering out the string
+   * Tracks event with all parent actionContext names to while filtering out the string
    * "action" (used in ExperimentVariable names but filtered out from event names).
    *
    * @param eventName Current event.
-   * @return Prefixed event name with all parent actions.
+   * @param value Value for event.
+   * @param info Info for event.
+   * @param params Dictionary of params for event.
    */
-  private String eventWithParentEventNames(String eventName) {
+  private void eventWithParentEventNames(String eventName, double value, String info,
+      Map<String, Object> params) {
     StringBuilder fullEventName = new StringBuilder();
     ActionContext context = this;
     List<ActionContext> parents = new ArrayList<>();
     while (context.parentContext != null) {
       parents.add(context);
       context = context.parentContext;
     }
+    Map<String, StringBuilder> mapOfMessages = new HashMap<>();
     for (int i = parents.size() - 1; i >= -1; i--) {
-      if (fullEventName.length() > 0) {
-        fullEventName.append(' ');
-      }
+      String messageIdForTrack = messageId;
       String actionName;
+      context = this;
       if (i > -1) {
-        actionName = parents.get(i).key;
+        context = parents.get(i);
+        actionName = context.key;
+        if (context.chainedToExisting) {
+          messageIdForTrack = context.parentContext.messageId;
+        }
       } else {
         actionName = eventName;
       }
-      if (actionName == null) {
+      if (actionName == null || TextUtils.isEmpty(actionName)) {
```

Comment:
`TextUtils.isEmpty(actionName)` checks for null before checking for length, so `actionName == null` is redundant